table/metadata.go

```diff
@@ -0,0 +1,401 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package table
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/apache/iceberg-go"
+
+	"github.com/google/uuid"
+)
+
+// Metadata for an iceberg table as specified in the Iceberg spec
+//
+// https://iceberg.apache.org/spec/#iceberg-table-spec
+type Metadata interface {
+	// Version indicates the version of this metadata, 1 for V1, 2 for V2, etc.
+	Version() int
+	// TableUUID returns a UUID that identifies the table, generated when the
+	// table is created. Implementations must throw an exception if a table's
+	// UUID does not match the expected UUID after refreshing metadata.
+	TableUUID() uuid.UUID
+	// Location is the table's base location. This is used by writers to determine
+	// where to store data files, manifest files, and table metadata files.
+	Location() string
+	// LastUpdatedMillis is the timestamp in milliseconds from the unix epoch when
+	// the table was last updated. Each table metadata file should update this
+	// field just before writing.
+	LastUpdatedMillis() int64
+	// LastColumn returns the highest assigned column ID for the table.
+	// This is used to ensure fields are always assigned an unused ID when
+	// evolving schemas.
+	LastColumn() int
+	// Schemas returns the list of schemas, stored as objects with their
+	// schema-id.
+	Schemas() []*iceberg.Schema
+	// CurrentSchema returns the table's current schema.
+	CurrentSchema() *iceberg.Schema
+	// PartitionSpecs returns the list of all partition specs in the table.
+	PartitionSpecs() []iceberg.PartitionSpec
+	// PartitionSpec returns the current partition spec that the table is using.
+	PartitionSpec() iceberg.PartitionSpec
+	// DefaultPartitionSpec is the ID of the current spec that writers should
+	// use by default.
+	DefaultPartitionSpec() int
+	// LastPartitionSpecID is the highest assigned partition field ID across
+	// all partition specs for the table. This is used to ensure partition
+	// fields are always assigned an unused ID when evolving specs.
+	LastPartitionSpecID() *int
+	// Snapshots returns the list of valid snapshots. Valid snapshots are
+	// snapshots for which all data files exist in the file system. A data
+	// file must not be deleted from the file system until the last snapshot
+	// in which it was listed is garbage collected.
+	Snapshots() []Snapshot
+	// SnapshotByID find and return a specific snapshot by its ID. Returns
+	// nil if the ID is not found in the list of snapshots.
+	SnapshotByID(int64) *Snapshot
+	// SnapshotByName searches the list of snapshots for a snapshot with a given
+	// ref name. Returns nil if there's no ref with this name for a snapshot.
+	SnapshotByName(name string) *Snapshot
```

Comment:
same here, I think this should be plural

Reply:
Same as above, it's singular because it only returns 1 snapshot, not multiple. Just to clarify my understanding of iceberg a bit, is it possible / legal for there to be multiple snapshots with the same ref name? or will it always be exactly 1 or 0 snapshots?

Reply:
it should be either 0 or 1 snapshot that is returned. In my head I was thinking about `snapshotsById` that we use in the Java impl, so having singular here is fine I think